pdns/recursordist/syncres.cc

```diff
@@ -4280,18 +4290,28 @@ void SyncRes::sanitizeRecords(const std::string& prefix, LWResult& lwr, const DN
            are sending such responses */
         if (rec->d_type != QType::CNAME || qname != rec->d_name) {
           LOG(prefix << qname << ": Removing record '" << rec->toString() << "' in the ANSWER section without the AA bit set received from " << auth << endl);
-          rec = lwr.d_records.erase(rec);
+          skipvec[counter] = true;
+          ++skipCount;
           continue;
         }
       }
       // Disallow answer records not answering the QType requested. ANY, CNAME, DNAME, RRSIG complicate matters here
       if (qtype != QType::ANY && rec->d_type != qtype.getCode() && !isRedirection(rec->d_type) && rec->d_type != QType::RRSIG) {
         LOG(prefix << qname << ": Removing irrelevant record '" << rec->toString() << "' in the ANSWER section received from " << auth << endl);
-        rec = lwr.d_records.erase(rec);
+        skipvec[counter] = true;
+        ++skipCount;
         continue;
       }
 
       haveAnswers = true;
+      if (rec->d_type == QType::CNAME) {
+        if (auto cnametarget = getRR<CNAMERecordContent>(*rec); cnametarget != nullptr) {
+          allowedAnswerNames.insert(cnametarget->getTarget());
+        }
+      }
+      else if (rec->d_type == QType::DNAME) {
```

Comment:
Perhaps we could add a comment to explain that we will synthesize the `CNAME` later in the process, as I'm always unsure about what happens when we encounter a `DNAME`?